draft-ietf-masque-connect-ip.md

```diff
@@ -1271,6 +1271,58 @@ ADDRESS_ASSIGN capsule is parsed. This allows modifications to address
 assignement to operate atomically. Similarly, extensions that modify routing
 SHOULD behave similarly with regards to the ROUTE_ADVERTISEMENT capsule.
 
+# Performance Considerations
+
+Bursty traffic can often lead to temporally-correlated packet losses; in turn,
+this can lead to suboptimal responses from congestion controllers in protocols
+running inside the tunnel. To avoid this, endpoints SHOULD strive to avoid
+increasing burstiness of IP traffic; they SHOULD NOT queue packets in order to
+increase batching beyond the minimal amount required to take advantage of
+hardware offloads.
+
+When the protocol running inside the tunnel uses congestion control (e.g.,
+{{TCP}} or {{QUIC}}), the proxied traffic will incur at least two nested
+congestion controllers. The underlying HTTP connection MUST NOT disable
+congestion control unless it has an out-of-band way of knowing with absolute
+certainty that all of the inner packets belong to congestion-controlled
+connections.
+
+When the protocol running inside the tunnel uses loss recovery (e.g., {{TCP}}
+or {{QUIC}}), and the underlying HTTP connection runs over TCP, the proxied
+traffic will incur at least two nested loss recovery mechanisms. This can
+reduce performance as both can sometimes independently retransmit the same
+data. To avoid this, IP proxying SHOULD be performed over HTTP/3 to allow
+leveraging the QUIC DATAGRAM frame.
+
+## MTU Considerations
+
+When using HTTP/3 with the QUIC Datagram extension {{DGRAM}}, IP packets are
+transmitted in QUIC DATAGRAM frames. Since these frames cannot be fragmented,
+they can only carry packets up to a given length determined by the QUIC
+connection configuration and the Path MTU (PMTU). If an endpoint is using QUIC
+DATAGRAM frames and it attempts to route an IP packet through the tunnel that
+will not fit inside a QUIC DATAGRAM frame, the IP proxy SHOULD NOT send the IP
+packet in a DATAGRAM capsule, as that defeats the end-to-end unreliability
+characteristic that methods such as Datagram Packetization Layer PMTU Discovery
+(DPLPMTUD) depend on {{?DPLPMTUD=RFC8899}}. In this scenario, the endpoint
+SHOULD drop the IP packet and send an ICMP Packet Too Big message to the sender
+of the dropped packet; see {{Section 3.2 of ICMPv6}}.
+
+## ECN Considerations
+
+If a client or IP proxy with a connection containing an IP Proxying request
+stream disables congestion control, it MUST NOT signal Explicit Congestion
+Notification (ECN) {{!ECN=RFC3168}} support on that outer connection. That is,
+it MUST mark all IP headers with the Not-ECT codepoint. It MAY continue to
+report ECN feedback via QUIC ACK_ECN frames or the TCP ECE bit, as the peer may
+not have disabled congestion control.
```

Comment:
So I have spent a bit of time thinking about the ECN marking. Aspect and when back and reviewed https://www.ietf.org/archive/id/draft-ietf-tsvwg-ecn-encap-guidelines-17.html and https://datatracker.ietf.org/doc/draft-ietf-tsvwg-rfc6040update-shim/

So for reliable stream encapsulation as well as for datagram over a congestion controlled QUIC connection the tunnel ingress will have to CE mark or drop packets when the queue gets to big to send an end-to-end congestion signal. And that is correctly reflected here. 

Next is this mode of datagram transmitted packets where the QUIC connection does not congestion control them with the assumption that the traffic as it is IP would be congestion controlled or otherwise safe for Internet use. That has been argued before, see Section 4.3.2 of https://datatracker.ietf.org/doc/draft-ietf-intarea-tunnels/ for such an argument with references. However, when it comes to such a tunnel we can't apply for example Feed forward and up marking (https://www.ietf.org/archive/id/draft-ietf-tsvwg-ecn-encap-guidelines-17.html#name-feed-forward-and-up-mode-gu) without explicit negotiation that the receiving peer would actually perform RFC 6040 recombing of the outer and the inner ECN field. In addition one would likely have to say that in this mode that one should never coalescing of multiple DATAGRAM frames in the same QUIC packets. 

There is also the accounting problem for ECT fields with QUIC ACKs as they only contain counters, not explicit ack vectors for ECN fields. Thus, the tunnel ingress sender can't separate the non-CC controlled packets from the one that is congestion controlled in regards to ECN markings. Thus, the receiver would likely have to be in on which QUIC packets are not counted and should not be reported so that they do not increase the counter, or have a counter for how many packets should be ignored from CC perspective.  

So for now I think this document do specifies what is possible to do without causing issues. It is not what is desired but dealing with ECN markings will require MASQUE extensions for signalling as well as methods for handling traffic appropriately when doing the desired feed forward and up tunnel. 

@bbriscoe do you agree with my conclusions here?

Reply:
There's a whole world of potential complexity we could pull in here, but I'm in agreement with Bob that we shouldn't specify things that won't get implemented. So I'd propose to stick with what we have now, and defer to an extension to specify this once someone has implemented it

Reply:
@DavidSchinazi  think we are in basic agreement on the way forward. Still I think the ECN text needs adjustment see comment above. And possible from the below to. 

To close down on a few things Bob said in his reply. 
From @bbriscoe 
> [BB] IIUC {Note 3}, I agree.
> BUT, I don't agree with either premise. I.e. I don't agree that either
> i) reliable stream encap or ii) datagram over a congestion controlled
> QUIC connection are sensible.
>
> {Note 3}: My parser is having problems with the above sentence. I think
> it means
> "...the tunnel ingress will have to CE mark or drop packets to send an
> end-to-end congestion signal when its queue gets too big."
> Otherwise, as written, it seems to mean the following, which I assume
> wasn't intended:
> " the tunnel ingress will have to CE mark or drop packets when the queue
> gets so big that it cannot send an end-to-end congestion signal"

So for a tunnel to achieve maximum performance I agree with what you write. 

So fully reliable encap will only occur when someone applies this mechanism over HTTP 1.1. and HTTP/2 or HTTP/3 without QUIC datagram. The main intention here is non-reliable encap using HTTP/3 with datagrams to avoid that issue. 

When it comes to congestion controlled aspekt of QUIC that is a part I wanted to ensure that we try to support as best we can. So the proposed text, possibly with some amendments would say that it would be possible to do it, but we are not really giving details for how to exclude QUIC packets with datagrams with encapsulated IP packets from the congestion control. We have some grey zones that maybe we should be more specific in regards to coalescing inside a QUIC packet and to not do that for packets. I would argue that the simplest way forward here is really to just send one QUIC packet per IP packet one tunnels independent of size. Including a non ack eliciting QUIC frame is likely also fine if it fits.

However, doing the above has the result that ECN will have to be either of two things.

1) Disabled by setting the ECN field to no-ECT for all outer QUIC packets that carry tunneled IP packets in datagrams outside of congestion control. 

2) Build an extension that negotiates the capability to do 6040 combining of ECN fields from outer to inner on the egress and possibly have an header/parameter in the extended connect request to indicate that what streams are handled outside of congestion control. As well as specify how do deal with coalescing etc. 

To be clear, I am only proposing 1) for now. It is the minimal and correct but not optimal. 

2) will have to be done later. That should go into the necessary complexities of handling L4S etc also. 

So even with 1) I think there are some small aspects we could consider to clarify in this document around the IP tunnelling in regards to Bob's comments.

A) Handling of traffic class (DSCP). With a congestion controlled QUIC connection you will only have one traffic behaviour from the tunnel, if you want to take it into account you need one tunnel per traffic class. With non-congestion controlled you can mark/translate from the inner to the outer field at ingress. 

B) Say anything about if coalescing of QUIC Datagram Frames are possible or not. Currently we are totally silent on this.